### ISSUE TITLE
creating `experimental` package for models or acquisitions that are not yet thoroughly tested? 

### ISSUE THREAD

**Issue by hstojic** [secondmind-labs/trieste#345]
What to do with GPflux and Keras models support, that is currently under construction? 

These are big PRs and its costly drag them along for a long time until we verify with experiments that they work reasonably well in BO setting. Because of their size they are costly to keep them up to date with `develop`.

Potential solutions:
1. Merge them in the `develop` as a part of the `model` package. These models should be tested (unit and integration tests) to a similar extent as GPflow models, but we would need to clearly warn users there is little literature on using these models in BO setting (i.e. not enough evidence they produce great results) and it's more risky to use them. 
2. Special `trieste/experimental` package where we would place all experimental models (possibly acquisition functions as well). Folder structure should mirror the rest of `trieste`, e.g. `trieste/experimental/models/gpflux/` and we should have some rules in place for transitioning from experimental to core. Advantage of this option warns users more explicitly that these modules are experimental, perhaps not tested as extensively, interfaces could break etc.

It seems that Option 2 is favoured, but perhaps we would benefit from further discussion of costs/benefits of each option - any thoughts?
How do other projects deal with support of experimental code?

**Comment by apaleyes** [comment]
I don't know the best answer here, but have a couple of thoughts.

Personally I am indeed in favor of option 2. It reminds a lot of Python's `from __future__ import ...` or [tf.experimental](https://www.tensorflow.org/api_docs/python/tf/experimental), this is also something we did in Emukit. It's quite convenient to get going with. It is also very obvious for the users, which helps with breaking backwards compatibility. The flip side though is that we would need a pretty clear process/criteria in place for graduating features from experimental namespace to the core, and a regular review. That is not always easy to do, especially in fast moving lively open source project. By the way of the same examples, Python seems to be able to manage this to some extent, while in Emukit we only ever graduated one feature. I don't know how well that works out for TF. There is also a big incentive to just throw any half-baked implementation under "experimental". Normally done on a temp basis, such things often tend to be frustratingly permanent. 

Option 1 is a bit better for maintenance, because if the feature is added in the core level at the start, it is always showing up in tests, coverage reports, etc. So it's much harder to ignore, and thus much more likely to eventually mature. On the other hand, it's much more obscure for the user, so there is a bigger chance users may rely on it, and thus it's harder to make breaking changes. Given how likely changes are to occur in the PRs you mentioned, i think that's a worse of two options.

I suppose there is also option 3, which is creating separate addon-like Python packages. E.g. "secondmind-labs/trieste-gpflux" or "secondmind-labs/trieste-keras". This is the same "experimental" idea, just taken to a bit higher level. The advantage here is that it's very clear that this is a separate effort, it can evolve separately, and can be archived at any point if we see the idea is not worth pursuing. It can also be eventually merged into Trieste. What's more, there is no need to keep up with Trieste's develop updates, now these can just use Trieste as a dependency, just as any other package would. Clear downside is all the maintenance overhead that comes with a Python package. And of course all of this assumes this is technically possible to separate those PRs from Trieste's internals, but I am not familiar with technical details of these PRs enough yet to make that call.

It is important to realize that none of these is solving the tech debt of keeping the PRs in sync with develop branch, they are all just delaying it and/or moving it elsewhere. Given this understanding, Option 2 feels as the best trade-off, but only after setting up a process for things entering the "experimental" namespace and graduating from it.

**Comment by sebastianober** [comment]
Thanks for opening this issue Hrvoje. I don't particularly have a strong opinion on this, although I would probably be leaning towards option 2 at the moment. I agree with Andrei that if we go this route we would need a clear process for graduating additions to the experimental package to the main package, and that we need a way of discouraging half-baked ideas that likely won't make it to the main package being thrown in. I also agree with Andrei that this doesn't really solve the tech debt of keeping PRs in sync with the develop branch, and nor does it really solve the problem that these PRs are big and will require a lot of effort to review properly when it comes time to merge them in with the main branch.

Overall if Hrvoje wants to make a decision on this I'm happy to follow whatever that would be.

**Comment by hstojic** [comment]
Thank everyone for your comments (both here and offline).

Option 4 appeared in the meantime: Include the new models in the core code, but try to keep the support at the minimum, by providing the interfaces (and some basic model?), and move the remainder to the experimental sub-packages.

After some further discussions, I would favour something close to Option 1 or Option 4. Points for including these branches into the core `models` package (most of it, that is):

- the neural network and deep GPs code is probably already well tested (close to what we have for GPflow, with GPflux being closer to it)
- in terms of breaking changes, design of interfaces closely followed those for GPflow, probably we won't be changing them more than we would change GPflow code

Essentially, they are past the point of really being experimental in the above sense. Some parts of it could potentially go to experimental (e.g. less tested models, `MonteCarloExpectedImprovement`). 

My original concern was with not having abundant evidence that these models work well in Bayesian optimisation setting, i.e. there is not a lot of research literature. But after some further thought, there is certainly some literature, and its not easy to draw a line here, what would be good enough evidence? That would delay things too much and this code probably would never be merged in. Then, there is the "availability" effect, the fact that it would be easy to test these models out in the toolbox could spur additional research and models would be explored much faster in wider variety of settings. Hence, I would dispense with the original concern and go for it.

**Comment by vpicheny** [comment]
I'm not sure what option 4 is exactly but it looks like something I could prefer. I think that skeleton interfaces with keras and gpflux should definitely be part of trieste. The experimental part are specific models. For gpflow, we decided to include interfaces for most models, along with decent default solutions. We probably don't want to do this for NNs, not sure about DGPs (if we can have something that works correctly then it's a great selling argument).

**Comment by hstojic** [comment]
Thanks everyone for opinions. For now we went with Option 1 for above reasons, but including basic models only to minimize support for the time being, and leaving other models in branches for the sake of record keeping, instead of complicating things with `experimental`
